File: ProjectEuler_FSharp/Primes.fs
module ProjectEuler_FSharp.Primes

open System
open System.Collections.Generic

let UpTo maxValue =
    let isComposite : bool array = Array.zeroCreate (maxValue + 1)
    seq {
        for i = 2 to maxValue do
        if not isComposite.[i] then
            yield i
            for j in (2 * i)..i..maxValue do
                isComposite.[j] <- true }

let UpToBig maxValue =
    let addNextPrime (previousPrimes: List<int64>, candidate) =
        let primeLimit = int64(Math.Sqrt(float(candidate))) + 1L;
        let candidateIsPrime =
            previousPrimes
            |> Seq.takeWhile (fun p -> p < primeLimit)
            |> Seq.exists (fun p -> candidate % p = 0L)
            |> not
        if candidateIsPrime then
            previousPrimes.Add(candidate)
        candidateIsPrime && candidate <= maxValue
    
    let primes = new List<int64>()
    seq {
        yield 2L
        yield 3L
        for i in 6L..6L..(maxValue + 10L) do
            if addNextPrime(primes, i - 1L) then
                yield i - 1L
            if addNextPrime(primes, i + 1L) then
                yield i + 1L
    }

File: ProjectEuler_CSharp/Primes.cs
using System;
using System.Collections.Generic;

namespace ProjectEuler_CSharp
{
    public static class Primes
    {
        public static IEnumerable<int> UpTo(int maxValue)
        {
            var isComposite = new bool[maxValue + 1];
            for (int i = 2; i <= maxValue; i++) {
                if (!isComposite[i]) {
                    yield return i;
                    for (int j = 2 * i; j <= maxValue; j += i)
                        isComposite[j] = true;
                }
            }
        }

        public static IEnumerable<long> UpTo(long maxValue)
        {
            var primes = new List<long> { 5, 7 };
            yield return 2;
            yield return 3;
            foreach (var p in primes) yield return p;

            bool testCandidate(long candidate)
            {
                long squareRoot = (long)Math.Sqrt(candidate) + 1;
                foreach (var prime in primes) {
                    if (candidate % prime == 0)
                        return false;
                    else if (prime > squareRoot)
                        return true;
                }
                return true;
            }

            long nextCandidate;
            for (long candidate = 12; candidate <= maxValue + 10; candidate += 6) {
                nextCandidate = candidate - 1;
                if (nextCandidate > maxValue)
                    break;
                if (testCandidate(nextCandidate)) {
                    primes.Add(nextCandidate);
                    yield return nextCandidate;
                }

                nextCandidate = candidate + 1;
                if (nextCandidate > maxValue)
                    break;
                if (testCandidate(nextCandidate)) {
                    primes.Add(nextCandidate);
                    yield return nextCandidate;
                }
            }
        }

    }
}
